scylla/src/history.rs

```diff
@@ -432,3 +432,649 @@ fn write_fiber_attempts(fiber: &FiberHistory, f: &mut std::fmt::Formatter<'_>) -
+
+        for query in &mut history.queries {
+            for fiber in std::iter::once(&mut query.non_speculative_fiber)
+                .chain(query.speculative_fibers.iter_mut())
+            {
+                for attempt in &mut fiber.attempts {
+                    attempt.node_addr = the_node;
+                }
+            }
+        }
+
+        history
+    }
+
+    // Set a single error message for all DbErrors within StructuredHistory.
+    // The error message changes between Scylla/Cassandra/their versions.
+    // Setting it to one value makes it possible to run tests consistently.
+    fn set_one_db_error_message(mut history: StructuredHistory) -> StructuredHistory {
+        let set_msg = |err: &mut QueryError| {
+            if let QueryError::DbError(_, msg) = err {
+                *msg = "Error message from database".to_string();
+            }
+        };
+
+        for query in &mut history.queries {
+            if let Some(QueryHistoryResult::Error(_, err)) = &mut query.result {
+                set_msg(err);
+            }
+            for fiber in std::iter::once(&mut query.non_speculative_fiber)
+                .chain(query.speculative_fibers.iter_mut())
+            {
+                for attempt in &mut fiber.attempts {
+                    if let Some(AttemptResult::Error(_, err, _)) = &mut attempt.result {
+                        set_msg(err);
+                    }
+                }
+            }
+        }
+
+        history
+    }
+
+    fn node1_addr() -> SocketAddr {
+        SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)), 19042)
+    }
+
+    fn node2_addr() -> SocketAddr {
+        SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 2)), 19042)
+    }
+
+    fn node3_addr() -> SocketAddr {
+        SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 3)), 19042)
+    }
+
+    fn timeout_error() -> QueryError {
+        QueryError::TimeoutError
+    }
+
+    fn unavailable_error() -> QueryError {
+        QueryError::DbError(
+            DbError::Unavailable {
+                consistency: LegacyConsistency::Regular(Consistency::Quorum),
+                required: 2,
+                alive: 1,
+            },
+            "Not enough nodes to satisfy consistency".to_string(),
+        )
+    }
+
+    fn no_stream_id_error() -> QueryError {
+        QueryError::UnableToAllocStreamId
+    }
+
+    #[test]
+    fn empty_history() {
+        let history_collector = HistoryCollector::new();
+        let history: StructuredHistory = history_collector.clone_structured_history();
+
+        assert!(history.queries.is_empty());
+
+        let displayed = "Queries History:
+";
+        assert_eq!(displayed, format!("{}", history));
+    }
+
+    #[test]
+    fn empty_query() {
+        let history_collector = HistoryCollector::new();
+
+        let _query_id: QueryId = history_collector.log_query_start();
+
+        let history: StructuredHistory = history_collector.clone_structured_history();
```

Comment:
Why do you prefer to clone instead of taking the structured history?

Reply:
I didn't want to think whether taking is ok or not, so I just cloned stuff. Those are tests anyway.